File: clients/rust/src/instructions/refresh_kamino_reserve.rs
use solana_instruction::{AccountMeta, Instruction};
use solana_pubkey::Pubkey;

use crate::{integrations::kamino::derive_anchor_discriminator, KAMINO_LEND_PROGRAM_ID};

pub fn create_refresh_kamino_reserve_instruction(
    reserve: &Pubkey,
    market: &Pubkey,
    scope_prices: &Pubkey,
) -> Instruction {
    let data = derive_anchor_discriminator("global", "refresh_reserve");

    Instruction {
        program_id: KAMINO_LEND_PROGRAM_ID,
        accounts: vec![
            AccountMeta {
                pubkey: *reserve,
                is_signer: false,
                is_writable: true,
            },
            AccountMeta {
                pubkey: *market,
                is_signer: false,
                is_writable: false,
            },
            // pyth oracle
            AccountMeta {
                pubkey: KAMINO_LEND_PROGRAM_ID,
                is_signer: false,
                is_writable: false,
            },
            // switchboard_price_oracle
            AccountMeta {
                pubkey: KAMINO_LEND_PROGRAM_ID,
                is_signer: false,
                is_writable: false,
            },
            // switchboard_twap_oracle
            AccountMeta {
                pubkey: KAMINO_LEND_PROGRAM_ID,
                is_signer: false,
                is_writable: false,
            },
            // scope_prices
            AccountMeta {
                pubkey: *scope_prices,
                is_signer: false,
                is_writable: false,
            },
        ],
        data: data.to_vec(),
    }
}

File: clients/ts/src/instructions/refreshKaminoReserve.ts
import { Address, AccountMeta, Instruction, AccountRole } from "@solana/kit";
import { anchorDiscriminator } from "../integrations";

const KAMINO_LEND_PROGRAM_ID =
  "KLend2g3cP87fffoy8q1mQqGKjrxjC8boSyAYavgmjD" as Address;

/**
 * Instruction generation for refreshing a Kamino reserve
 */
export function createRefreshKaminoReserveInstruction(
  reserve: Address,
  market: Address,
  scopePrices: Address
): Instruction {
  const data = anchorDiscriminator("global", "refresh_reserve");

  const accounts: AccountMeta[] = [
    {
      address: reserve,
      role: AccountRole.WRITABLE,
    },
    {
      address: market,
      role: AccountRole.READONLY,
    },
    // pyth oracle
    {
      address: KAMINO_LEND_PROGRAM_ID,
      role: AccountRole.READONLY,
    },
    // switchboard_price_oracle
    {
      address: KAMINO_LEND_PROGRAM_ID,
      role: AccountRole.READONLY,
    },
    // switchboard_twap_oracle
    {
      address: KAMINO_LEND_PROGRAM_ID,
      role: AccountRole.READONLY,
    },
    // scope_prices
    {
      address: scopePrices,
      role: AccountRole.READONLY,
    },
  ];

  return {
    programAddress: KAMINO_LEND_PROGRAM_ID,
    accounts,
    data,
  };
}

